dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java

```diff
@@ -266,10 +266,39 @@ private void writeField(FieldDef fd, Object fieldValue) throws Exception
         {
             case BINARY:
             {
-                byte[] data = fieldValue==null?new byte[0]:(byte[]) fieldValue;
-                long dataSize = data.length;
-                writeUnsigned(dataSize);
-                writeByteArray(data);
+                if (fd.isFixed())
+                {
+                    byte[] data = fieldValue==null?new byte[0]:(byte[]) fieldValue;
+                    int bytesToWrite = data.length;
+                    if (bytesToWrite > fd.getDataLen())
+                    {
+                        bytesToWrite = (int) fd.getDataLen();
+                    }
+
+                    writeByteArray(data, 0, bytesToWrite);
+
+                    // Fill in remaining length with 0
+                    int numFillBytes = (int) fd.getDataLen() - bytesToWrite;
+                    while (numFillBytes > 0)
+                    {
+                        int fillLength = numFillBytes;
+                        if (fillLength > SCRATCH_BUFFER_SIZE)
+                        {
+                            fillLength = SCRATCH_BUFFER_SIZE;
+                        }
+
+                        Arrays.fill(scratchBuffer, 0, fillLength, (byte) '\0');
+                        writeByteArray(scratchBuffer, 0, fillLength);
+                        numFillBytes -= fillLength;
+                    }
+                }
+                else
+                {
+                    byte[] data = fieldValue==null?new byte[0]:(byte[]) fieldValue;
```

Comment:
minor but this line is repeated, if it doesn't affect maintainability lets move out of the conditional